### ISSUE TITLE
Joins support subqueries, not just table names

### ISSUE THREAD

**Issue by osdiab** [knex/knex#5838]
The docs don't clarify in the examples that the first argument in each `join` method, `table`, can actually be a subquery expressed as a function.

Examples here for `leftJoin`: https://stackoverflow.com/questions/48098061/knex-js-subqueries-on-mysql-left-join

I just inherited a codebase that uses this behavior, I would love it if the documentation, and correspondingly the typescript types, reflect this reality.

Usage example:

```
> knex.select('*').from('users').join(function() { this.select('*').from('accounts') }, 'a', '=', 'b').toSQL().sql
'select * from `users` inner join (select * from `accounts`) on `a` = `b`'
```

**Comment by mercmobily** [comment]
You're absolutely right that the join documentation doesn't clearly show that the `table` parameter can be a subquery expressed as a function, a QueryBuilder instance, or even `knex.raw()`. The current documentation only shows string table names, which is incomplete.

As elhigu confirmed in the comments, you can pass:
1. A callback function: `join(function() { this.select('*').from('accounts') }, ...)`
2. A QueryBuilder instance
3. A `knex.raw()` expression

Your example demonstrates this perfectly:
```javascript
knex.select('*').from('users')
  .join(function() {
    this.select('*').from('accounts')
  }, 'a', '=', 'b')
// Produces: select * from `users` inner join (select * from `accounts`) on `a` = `b`
```

However, as elhigu noted, there's an inconsistency: callback functions and QueryBuilder instances handle parentheses differently, which should also be documented. The join builder accepts a function, a QueryBuilder instance, or a `knex.raw()` for the `table` operand and compiles them via the formatter (lib/query/querybuilder.js:318–341; lib/formatter/wrappingFormatter.js:41–88, 90–133).

The documentation should be updated to include:
1. Examples showing subquery joins using callbacks
2. Examples using QueryBuilder instances
3. Examples using knex.raw()
4. A note about the parentheses inconsistency
5. Links to the StackOverflow answer and related discussions

As osdiab mentioned, the TypeScript types should also be updated to reflect that the first parameter can be more than just a string.

This is a valuable documentation improvement that would help developers understand the full capabilities of the join methods. A quick `toSQL()` of your example yields `select * from "users" inner join (select * from "accounts") on "a" = "b"`.
